blas/unit_test/Test_Blas2_syr2.hpp

```diff
@@ -0,0 +1,1970 @@
+  // Step 3 of 7: populate h_vanilla
+  // ********************************************************************
+  view_stride_adapter<_ViewTypeExpected, true> h_vanilla(
+      "vanilla = A + alpha * x * x^{t,h}", _M, _N);
+#ifdef HAVE_KOKKOSKERNELS_DEBUG
+  KOKKOS_IMPL_DO_NOT_USE_PRINTF(
+      "In Test_Blas2_syr2.hpp, computing vanilla A with alpha type = %s\n",
+      typeid(alpha).name());
+#endif
+  this->populateVanillaValues(alpha, x.h_view, y.h_view, A.h_view,
+                              h_vanilla.d_view);
+
+  // ********************************************************************
+  // Step 4 of 7: use h_vanilla and h_expected as appropriate
+  // ********************************************************************
+  if (expectedResultIsKnown) {
+    // ******************************************************************
+    // Compare h_vanilla against h_expected
+    // ******************************************************************
+    this->compareVanillaAgainstExpected(alpha, h_vanilla.d_view,
+                                        h_expected.d_view);
+  } else {
+    // ******************************************************************
+    // Copy h_vanilla to h_expected
+    // ******************************************************************
+    Kokkos::deep_copy(h_expected.d_base, h_vanilla.d_base);
+  }
+
+  // ********************************************************************
+  // Step 5 of 7: test with 'non const x'
+  // ********************************************************************
+  view_stride_adapter<_ViewTypeA, false> org_A("Org_A", _M, _N);
+  Kokkos::deep_copy(org_A.d_base, A.d_base);
+  Kokkos::deep_copy(org_A.h_view, A.h_view);
+
+  if (test_x) {
+    this->callKkSyr2AndCompareAgainstExpected(alpha, x.d_view, y.d_view,
+                                              A.d_view, A.h_view,
+                                              h_expected.d_view, "non const x");
+
+    if ((_useAnalyticalResults == false) &&  // Just to save run time
+        (_kkGerShouldThrowException == false)) {
+      this->callKkGerAndCompareKkSyr2AgainstIt(alpha, x.d_view, y.d_view, org_A,
+                                               A.h_view, "non const x");
+    }
+  }
+
+  // ********************************************************************
+  // Step 6 of 7: test with const x
+  // ********************************************************************
+  if (test_cx) {
+    Kokkos::deep_copy(A.d_base, org_A.d_base);
+
+    this->callKkSyr2AndCompareAgainstExpected(
+        alpha, x.d_view_const, y.d_view_const, A.d_view, A.h_view,
+        h_expected.d_view, "const x");
+  }
+
+  // ********************************************************************
+  // Step 7 of 7: tests with invalid values on the first input parameter
+  // ********************************************************************
+  EXPECT_ANY_THROW(
+      KokkosBlas::syr2(".", "U", alpha, x.d_view, y.d_view, A.d_view))
```

Comment:
Probably not necessary here but in general we might want to fence after kernels calls since these are non-blocking

Reply:
Thanks for the hint.